### PR TITLE
Backport: Store legacy sidebars when switching to a block theme

### PR DESCRIPTION
**Pull request by Mamaduka**
Trac ticket: https://core.trac.wordpress.org/ticket/57531
Related Gutenberg PR: https://github.com/WordPress/gutenberg/pull/45509

## Testing instructions without package updates

1. Switch to the Twenty Twenty theme.
2. Confirm sidebars have at least one widget by going to Appearance > Widgets
3. Switch to any block theme. I was using Twenty Twenty-Three in my tests.
4. Make a get request to the sidebars endpoint. It requires auth, but the Application Password could be used for basic auth.
5. Confirm that the request loads the same sidebars registered for the Twenty Twenty theme.

**Example request made by the editor.**

```http
GET /wp-json/wp/v2/sidebars?context=edit&_fields=id,name,description,status,widgets
```

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by Mamaduka** [comment]
@hellofromtonya, I'm unsure what would be the right place to add PHPUnit tests for this change or the actual "unit" we want to test. In the plugin, the feature is covered by e2e tests.

File: src/wp-includes/theme.php [hunk 1]
```diff
@@ -737,7 +737,7 @@ function locale_stylesheet() {
  * @param string $stylesheet Stylesheet name.
  */
 function switch_theme( $stylesheet ) {
-	global $wp_theme_directories, $wp_customize, $sidebars_widgets;
+	global $wp_theme_directories, $wp_customize, $sidebars_widgets, $wp_registered_sidebars;
```
Comment (mukeshpanchal27) [inline src/wp-includes/theme.php:741]: Quick review: Document missing for the `$wp_registered_sidebars`

Comment (Mamaduka) [reply src/wp-includes/theme.php:741]: Resolved in [ba0fa42](https://github.com/WordPress/wordpress-develop/pull/3893/commits/ba0fa4294637c45d98ae2e1f7607904250e1cdc2).

File: src/wp-includes/widgets.php [hunk 2]
```diff
@@ -2105,3 +2105,29 @@ function wp_check_widget_editor_deps() {
 		}
 	}
 }
+
+/**
+ * Register the previous theme's sidebars for the block themes.
```
Comment (audrasjb) [inline src/wp-includes/widgets.php]: Nitpick: Use third-person forms in docblock shortdesc: "Register**s**".

**Review by audrasjb — COMMENTED** [review]
@Mamaduka I aded a small comment, but otherwise the logic looks good to me 👍

**Comment by Mamaduka** [comment]
Thank you, @audrasjb!

Do you have any suggestions about the PHPUnit tests - https://github.com/WordPress/wordpress-develop/pull/3893#issuecomment-1400432537?

**Comment by audrasjb** [comment]
Concerning unit tests, maybe try to verify that classic sidebars are stored when switching to a block theme?

**Comment by Mamaduka** [comment]
That makes sense, thank you.

I like the "classic sidebars" term and reference our other discussions from https://github.com/WordPress/wordpress-develop/pull/3902#discussion_r1085740716. Do you think I should change `legacy_sidebars` to `classic_sidebars`?

**Comment by audrasjb** [comment]
Yeah definitely! I think it would be easier to identify what is it: the system used before blocks started to handle WordPress content dataset :)

Yes it is legacy, but "legacy" can also be used for "legacy Classic" functions/classes/methods/concepts (= legacy-legacy stuff 🤪 ). Let's disambiguate when possible! :D

File: src/wp-includes/widgets.php (hunk 2 above)
Comment (Mamaduka) [reply src/wp-includes/widgets.php]: Done ✅

**Comment by Mamaduka** [comment]
* Fixed docblock.
* Updated the theme mod name.
* Added simple unit tests.

File: tests/phpunit/tests/widgets.php [hunk 3]
```diff
@@ -1364,4 +1364,24 @@ public function test_sidebar_guessing_one_menu_per_sidebar() {
...
+		global $wp_registered_sidebars;
+
+		$this->register_sidebars( array( 'sidebar-1', 'sidebar-2' ) );
+
+		switch_theme( 'block-theme' );
+
+		unregister_sidebar( 'sidebar-1' );
+		unregister_sidebar( 'sidebar-2' );
+
+		// Register stored classic sidebars for block themes..
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/widgets.php]:
```suggestion
		// Register stored classic sidebars for block themes.
```

Remove extra full stop.

**Comment by hellofromtonya** [comment]
>I like the "classic sidebars" term and reference our other discussions from https://github.com/WordPress/wordpress-develop/pull/3902#discussion_r1085740716. Do you think I should change legacy_sidebars to classic_sidebars?

I agree with @audrasjb that "classic" is preferred over "legacy" 👍 

@Mamaduka I'll look at the PHPUnit. Thank you for adding it!

Test Reports: I asked @ironprogrammer to do his magic in testing this PR.  If anyone has time, please test and add a test report too.

**Comment by hellofromtonya** [comment]
@ironprogrammer tested the PR and submitted a Test Report https://core.trac.wordpress.org/ticket/57531#comment:13. Works 👍

**Comment by hellofromtonya** [comment]
Currently reviewing for commit.

**Comment by hellofromtonya** [comment]
@Mamaduka I switched the way this PR is tested by:

* Adding a new test class for the new function `_wp_block_theme_register_classic_sidebars()`.
* Testing the "inactive" state in the REST API.

By doing so, it's now testing different paths for this specific change including:

* Does the sidebar register on theme switch after invoking  _wp_block_theme_register_classic_sidebars()?
* Does the REST Controller set the sidebar to `inactive`?
* Does the `switch_theme()` set the theme mod?

**Review by hellofromtonya — APPROVED** [review]
* Confirmed this PR includes the changes from the GB PR ✅ 
* It has a test report ✅ 
* It has unit/integration tests ✅ 

It's ready for commit (assuming the CI jobs pass this time 🤣 ) 👍

**Comment by hellofromtonya** [comment]
Committed via https://core.trac.wordpress.org/changeset/55200.

**Pull request closed without merging**